src/TRoom.cpp

```diff
@@ -703,95 +725,630 @@ void TRoom::restore( QDataStream & ifs, int roomID, int version )
+                                      .arg( id, 6, QLatin1Char( '0' ) )
+                                      .arg( it.key(), 6, QLatin1Char( '0' ) );
+                // If size is less than or equal to 0 then there is nothing to print!!!
+                /* This is wrong, must modify thing being iterated over via iterator:
+                 * other.remove( it.key(), it.value() );
+                 */
+                it.remove();
+                mpRoomDB->mpMap->postMessage( warnMsg );
+            }
+            else if( ! ( _cmd.startsWith('1') || _cmd.startsWith('0') ) ) {
+                QString _nc = it.value();
+                int _nk = it.key();
+                _nc.prepend('0');
+                // Old, prepatched special exit could not have a lock
+                /*
+                 * This is wrong, must modify thing being iterated over via iterator:
+                 * other.remove( it.key(), it.value() );
+                 * other.insertMulti( _nk, _nc );
+                 */
+                replacements.insert( _nk, _nc );
+                it.remove();
+                QString warnMsg = tr( "[ INFO ]  - In room Id:%1 patching {internal fixup} of (special) exit to\n"
+                                                  "%2, was: \"%3\" now: \"%4\"." )
+                                      .arg( id, 6, QLatin1Char( '0' ) )
+                                      .arg( _nk, 6, QLatin1Char( '0' ) )
+                                      .arg( _cmd )
+                                      .arg( _nc );
+                mpRoomDB->mpMap->postMessage( warnMsg );
+            }
+        }
+        // Now finished with (mutable) iterator, can re-insert changed things
+        if( ! replacements.isEmpty() ) {
+            other.unite( replacements );
+            // unite() is OK to use here as we have already removed the
+            // key/value pairs that are to be replaced (otherwise they'd be
+            // duplicated!)
+        }
     }
-    if( northeast != -1 && ! mpRoomDB->getRoom(northeast) )
-    {
-        qWarning("TRoom::auditExits() WARNING: roomID:%6i REMOVING invalid (general) exit: \"northeast\"", id);
-        northeast = -1;
+
+    // Now do the exit room ids if there is any remapping
+    if( ! roomRemapping.isEmpty() ) {
+        QMutableMapIterator<int, QString> it( other );
+        QMultiMap<int, QString> replacements;
+        while( it.hasNext() ) {
+            it.next();
+            int exitRoomId = it.key();
+            QString exitText = it.value();
+            QString exitName = exitText.mid( 1 );
+
+            if( roomRemapping.contains( exitRoomId ) ) {
+                QString auditKey = QStringLiteral( "audit.remapped_special_exit.%1" ).arg( exitName );
+                userData.insert( auditKey, QString::number( exitRoomId ) );
+                QString infoMsg = tr( "[ INFO ]  - In room with Id: %1 correcting special exit \"%2\" that\n"
```

Comment:
How come Id is capitalised everywhere? I know the variable name is capitalised, but in normal text this looks strange. Just 'id' or 'ID' will work.